### PR TITLE
fix: make `extend_docs` always persist new docstrings across imports

### PR DESCRIPTION
**Pull request by thorimur**
This PR ensures that the new docstring added by `extend_docs` is always persisted across imports.

The docstring extension maintains a `List (Name × String)` alongside a `NameMap String`; the `NameMap` is used for state, while the `List` is used for exports. `addDocString` calls out to a function which implicitly assumes that no docstring is being added twice, and simply prepends the new entry to the `List`, regardless of whether the name already appears in a pair there. A `qsort` with respect to the name which is used during serialization then might rearrange the entries. After import, when we conduct binary search through the entries associated with the module to retrieve the docstring, we might land on the wrong one.

See [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/extend_docs.20persistence.20issue)

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/ExtendDoc.lean [hunk 1]
```diff
@@ -26,18 +26,46 @@ At least one of `before` and `after` must appear, but either one of them is opti
...
+    (entries, map.insert declName docString)
+where
+  /-- Replace the `String` in the first pair that contains `declName`. (`declName` is assumed to
+  appear somewhere in the list.) -/
+  modifyEntry : List (Name × String) → List (Name × String)
+  | e@(n, _) :: entries =>
+    if n == declName then (n, docString) :: entries else e :: modifyEntry entries
+  | [] => []
+
+
```
Comment (adomani) [inline Mathlib/Tactic/ExtendDoc.lean:57]:
```suggestion

```
That's as far as I can give suggestions to this PR!  :smile:

Comment (adomani) [reply Mathlib/Tactic/ExtendDoc.lean]: I'm not sure about what happened: I selected the *two* line breaks and removed *one* in the suggestion.  Maybe, I should have selected *one* line break and removed it?

Comment (thorimur) [reply Mathlib/Tactic/ExtendDoc.lean]: The source of your suggestion looks sensible! Hmm, maybe GitHub trims whitespace at a point it shouldn't? Or something? Either way, I'm blaming this on a bug... 😁

**Comment by kmill** [comment]
I think I might try fixing this in core. I ran into a similar issue this week with another extension.

**Comment by kmill** [comment]
https://github.com/leanprover/lean4/pull/5911

**Comment by thorimur** [comment]
Although a slight misuse of the tag, I've added "blocked by core PR" for now—if that PR lands, I'll close this one. But if the core PR is expected to take too long to land and propagate to Mathlib (hopefully not!), maybe we ought to merge this one in the meantime. I'll let the maintainers decide. :)

**Comment by kmill** [comment]
The Lean PR was just merged and the next Lean release candidate will be made in the next couple days — shall we close this PR now?

**Pull request closed without merging**